diff --git a/src/Makefile b/src/Makefile
index f56465d..3314874 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -113,6 +113,9 @@ XCFLAGS=
 # Enable GC64 mode for x64.
 #XCFLAGS+= -DLUAJIT_ENABLE_GC64
 #
+# Enable UTF-8 script support.
+#XCFLAGS+= -DLUAJIT_ENABLE_UTF8SCRIPT
+#
 ##############################################################################
 
 ##############################################################################
@@ -491,6 +494,7 @@ LJCORE_O= lj_gc.o lj_err.o lj_char.o lj_bc.o lj_obj.o lj_buf.o \
	  lj_ctype.o lj_cdata.o lj_cconv.o lj_ccall.o lj_ccallback.o \
	  lj_carith.o lj_clib.o lj_cparse.o \
	  lj_lib.o lj_alloc.o lib_aux.o \
+	  compat-5.2.o lioutf8.o \
	  $(LJLIB_O) lib_init.o

 LJVMCORE_O= $(LJVM_O) $(LJCORE_O)
diff --git a/src/lioutf8.c b/src/lioutf8.c
index 8bfaca4..e339b8c 100755
--- a/src/lioutf8.c
+++ b/src/lioutf8.c
@@ -67,7 +67,7 @@ FILE* utf8_popen(const char *filename, const char *mode)
     free( wmode );
     return fp;
 #else
-    return popen(filename, mode, stream); 
+    return popen(filename, mode); 
 #endif
 }
 
diff --git a/src/lib_os.c b/src/lib_os.c
index f0d813e..6a999f8 100644
--- a/src/lib_os.c
+++ b/src/lib_os.c
@@ -108,7 +108,7 @@ LJLIB_CF(os_getenv)
 #if LJ_TARGET_CONSOLE
   lua_pushnil(L);
 #else
-#ifdef LJ_UTF8
+#if LJ_UTF8
   char *env = utf8_getenv(luaL_checkstring(L, 1));
   lua_pushstring(L, env);  /* if NULL push nil */
   if ( env != NULL ) free(env);
